rfc-174-consolidate-frontend-rendering-apps.md

```diff
@@ -0,0 +1,330 @@
+
+`helpers`
+- General utility methods for a view.
+
+Under `/lib`
+
+`tasks`
+- Rake tasks
+
+Use `lib` rather than `app/lib` for custom code that enhances the app but doesn’t fit into a directory in `app`
+
+Under `/config`
+
+`data`
+- YAML, JSON, CSV config files. Data files should be placed here rather than under `/lib/data`.
+
+`locales`
+- Static content like lists of contact information should ideally be stored in the locale files to promote localisation.
+
+Namespaces will be used to group functionality e.g. for StepBySteps, and resources in the routes file. This structure will be documented in README of the host app.
+Each application will be analysed before being consolidated to ensure redundant code is not being moved over.
+
+### How the proposal will be carried out
+This is obviously a very large piece of work that will take some time to complete. We have identified the following steps that will be needed.
+
+* Prepare `frontend` to be the host
+    * convert frontend tests to RSpec
+    * modernise the tests
+    * audit `frontend` and make changes to the frontend directory structure to match “new layout”
+    * remove redundant code from frontend (non-blocking)
+    * remove publishing_api tasks from frontend (non-blocking)
+* Merge `feedback` into `frontend`
+    * use as a test case and use that to inform our next steps
+* Write a roadmap for app consolidation
+    * will be handled separately
+    * might be simply the order in which applications will be merged into the host
+    * or could be some apps merged together before merging into the host
+    * or some routes merging ahead of others
+* Consider and work on expected issues (see below)
+* Work to merge remaining apps
+
+### Steps as part of a merge
+We anticipate that any merge will likely involve some or all of the following steps.
+
+* Identify any issues with existing code
+    * size these issues
+    * either fix issues or record them as tech debt - we need to avoid trying to fix everything and concentrate on progressing consolidation
+* Replicate the rendering from the candidate to the host
+    * preserve the git history of the candidate app - add a permalink to the files being consolidated in the commit in the host app.
+* Change content api
+    * change the rendering app for the document type in the publishing app.
+    * republish content
+    * republish any special routes (things that don’t have a publishing app, like the homepage) - any special routes should be moved into special route publisher (we have an outstanding tech debt card relating to this)
+* Optimise the frontend code
+    * remove duplication
+* Remove the routes from the candidate
+* [Retire the candidate] once all routes have been removed
+
+### Expected issues
+Merging all of our apps together is not without difficulties and we acknowledge that a single app will present problems that our multiple apps do not.
```

Comment:
Will we see any of the benefits described from having less frontend apps, or will we only get the benefits when we have a single app?

I'm just looking at the number of different controllers in government frontend and feeling a bit scared about all that coming into frontend! We will have to be so ninja with reviews to prevent a wild west (recently wept whilst wrapping my head around [all these services](https://github.com/alphagov/collections/tree/main/app/services) in collections that are doing basically the same thing)

Have we considered other ways of combining to get the number down to 2 (or 3)? Or do you not think that's a viable option?

Reply:
I think anything that reduces the number of apps even a little bit is a benefit as it reduces the maintenance burden, so 5 apps is better than 8 etc.

The benefit of the single app is being able to remove static and slimmer. 

We will have to be quite strict in our reviews and making sure everything is properly namespaced, and in the agreed area of the code. I think that having contributor guidelines that details what is finally decided on here should make that a bit easier as there will be something to point to when doing a review.

When you say "other ways of combining" what do you mean?